extras/module-node/duk_module_node.c

```diff
@@ -0,0 +1,263 @@
+	(void) duk_get_prop_string(ctx, -1, "\xffrequireCache");
+	duk_dup(ctx, -3);
+
+	/* [ ... module stash req_cache module ] */
+
+	(void) duk_get_prop_string(ctx, -1, "id");
+	duk_dup(ctx, -2);
+	duk_put_prop(ctx, -4);
+
+	duk_pop_3(ctx);  /* [ ... module ] */
+}
+
+static void duk__del_cached_module(duk_context *ctx, const char *id) {
+	duk_push_global_stash(ctx);
+	(void) duk_get_prop_string(ctx, -1, "\xffrequireCache");
+	duk_del_prop_string(ctx, -1, id);
+	duk_pop_2(ctx);
+}
+
+static duk_ret_t duk__handle_require(duk_context *ctx) {
+	/*
+	 *  Value stack handling here is a bit sloppy but should be correct.
+	 *  Since this is a Duktape/C call, call handling will clean up any extra
+	 *  garbage values for us.
+	 */
+
+	const char *id;
+	duk_ret_t nrets;
+	const char *parent_id;
+	duk_idx_t module_idx;
+	duk_idx_t stash_idx;
+	duk_int_t ret;
+
+	duk_push_global_stash(ctx);
+	stash_idx = duk_normalize_index(ctx, -1);
+
+	duk_push_current_function(ctx);
+	(void) duk_get_prop_string(ctx, -1, "\xffmoduleID");
+	parent_id = duk_require_string(ctx, -1);
+
+	/* [ id stash require parent_id ] */
+
+	id = duk_require_string(ctx, 0);
+
+	(void) duk_get_prop_string(ctx, stash_idx, "\xffmodResolve");
+	duk_dup(ctx, 0);   /* module ID */
+	duk_dup(ctx, -3);  /* parent ID */
+	duk_call(ctx, 2);
+	if (!duk_is_string(ctx, -1)) {
+		duk_error(ctx, DUK_ERR_API_ERROR, "error in module resolution");
+	}
+
+	/* [ ... stash ... resolved_id ] */
+
+	id = duk_require_string(ctx, -1);
+	if (duk__get_cached_module(ctx, id)) {
+		goto have_module;  /* use the cached module */
+	} else {
+		duk__push_module_object(ctx, id);
+		duk__put_cached_module(ctx);  /* module remains on stack */
+
+		module_idx = duk_normalize_index(ctx, -1);
+
+		/* [ ... stash ... resolved_id module ] */
+
+		(void) duk_get_prop_string(ctx, stash_idx, "\xffmodLoad");
+		duk_dup(ctx, -3);  /* resolved ID */
+		(void) duk_get_prop_string(ctx, module_idx, "exports");
+		duk_dup(ctx, module_idx);
+		ret = duk_pcall(ctx, 3);
+		if (ret != DUK_EXEC_SUCCESS) {
+			duk__del_cached_module(ctx, id);
+			duk_throw(ctx);  /* rethrow */
+		}
+
+		/* [ ... module source? ] */
+
+		if (!duk_is_undefined(ctx, -1) && !duk_is_string(ctx, -1)) {
+			duk__del_cached_module(ctx, id);
+			duk_error(ctx, DUK_ERR_API_ERROR, "error in module loading");
+		}
+
+		if (duk_is_string(ctx, -1)) {  /* if we have source */
+			if (duk__eval_module_source(ctx) != DUK_EXEC_SUCCESS) {
+				duk__del_cached_module(ctx, id);
+				duk_throw(ctx);  /* rethrow */
+			}
+		}
+	}
+
+	/* fall through */
+
+ have_module:
+	/* [ ... module ] */
+
+	(void) duk_get_prop_string(ctx, -1, "exports");
+	return 1;
+}
+
+static void duk__push_require_function(duk_context *ctx, const char *id) {
+	duk_push_c_function(ctx, duk__handle_require, 1);
+	duk_push_string(ctx, "name");
+	duk_push_string(ctx, "require");
+	duk_def_prop(ctx, -3, DUK_DEFPROP_HAVE_VALUE);
+	duk_push_string(ctx, id);
+	duk_put_prop_string(ctx, -2, "\xffmoduleID");
```

Comment:
There's not really a very established convention for that but I've tended to capitalize only the first letter of an acronym (`requireId`). Similarly I'd probably favor `encodeJx()` over `encodeJX()`. It doesn't really matter though because it's entirely internal to the module loader.